### DISCUSSION TITLE
Control of the eraser

### DISCUSSION THREAD

**Discussion by francoisschwarzentruber** [tableaunoir/tableaunoir#110] [Q&A]
Here is a place to discuss the behavior of the eraser. I really like the idea of having an eraser where you do not choose its size, but Tableaunoir "understands" the size you need.
However, now, some students complained that the eraser is... too small ;).
If you have ideas you want to share,  do not hesitate!

**Comment by tekhnee** [comment]
This is one of the more innovative features of _tableaunoir_. Great idea! There are two potential problems with the current implementation:

First, to get a large eraser, one needs to shake the cursor too abruptly (or across large distances), risking accidental erasures. 

Second, the kinematic logic governing the eraser size doesn't seem to me intuitive enough. Is size a function of eraser speed alone? A function of "work" (distance covered and acceleration)?… The behavior is somewhat unpredictable. But maybe I just haven't gotten the hang of it yet.

**Comment by francoisschwarzentruber** [comment]
Thank you. I agree with your two points. Maybe not choosing the size is just an impossible dream. I have to think a bit about it.

**Comment by deleporte** [comment]
Hi. I haven't used tableaunoir a lot, but in my recent experience, the degree of shaking needed to increase the eraser size is too important. In fact, I was not even aware of this feature before reading this discussion, even though I had tried to erase things with the (teeny tiny) eraser.

In my opinion
1. I see no reason whatsoever for the eraser width to be as small as the pencil width (at least for the default pencil size). 
2. The largest eraser width is good, but it takes too long to reach it; I'm counting four whole seconds while shaking a lot ! I'd be happy if it could be less than a second.

Sorry for the negative comment ; the rest of tableaunoir is great !

**Reply by francoisschwarzentruber** [reply]
Hi, thank you for your message. I think I solved your point 2, making your point 1 not so important any more. I really love your test "please, the big eraser should be available in less than one second by shaking".

1. Well, sometimes, we need to erase very very small part of the board. So I leave it as it now.
2. I agree and it has been solved. My big mistake was that I did not take into account the distance when augmenting the "temperature" of the eraser. (the warmer the eraser is, the more it is close to increase its size)

I think the eraser is easier to use now.
-Having the big eraser is easy (less than one second!) 
- Having control of a very precise eraser is also easy: just be careful to the color of the eraser that indicates its "temperature".

Your comment is not negative at all. It is very positive and constructive.

**Reply by deleporte** [reply]
Thanks! I'm eager to test it at the next release!

**Reply by francoisschwarzentruber** [reply]
https://tableaunoir.github.io/ is already up to date!

**Comment by francoisschwarzentruber** [comment]
What do you think of having a temperature bar for the eraser that indicates precisely its temperature? This bar may replace the toolbar when the stylus/mouse is down when erasing.

![image](https://user-images.githubusercontent.com/43071857/115712649-4e97bb80-a375-11eb-94dd-b06445d9eb26.png)

The idea is to:
- explain the user how the eraser works without letting the user guess how it works
- having a good control during precise erasing

The drawback may be that:
- the replacement of the toolbar may be distracting
- the temperature bar may be distracting
- it may feel as a gadget

What do you think?

**Reply by deleporte** [reply]
With the current parameters, if I want to erase a large part and just start slashing, the eraser becomes large quickly so I'm not sure if it's necessary to indicate that the eraser width changes, people are just going to experience it unless they insist on never dragging the mouse when erasing.

I'm not convinced the toolbar has reached its final form (e.g. place navigational tools at the lower right corner ?). 

If I can make a wish, it would be a thickness button, next to the colour button: 4 or 5 thicknesses (and supplementary parameters in the settings menu). The colour button should be greyed out when the eraser is selected, and the user could choose between the default adaptative thickness eraser and 4 or 5 thicknesses.

If I can make a *second* wish, it would be a rectangle eraser: click and drag to form a rectangle, erases everything inside. Very handy when giving a lecture!

**Reply by francoisschwarzentruber** [reply]
Thank you for your suggestions!
OK, so the eraser temperature bar is a gadget! I tend to agree.

Place navigational tools at the lower right corner ? => do you mean the lower right corner of the screen? I agree that we may confuse them with the "Undo/redo" buttons.

**Thickness button**
Maybe you are right, but not sure. As it is *not* a drawing software, I am not completely convinced by the thickness button. The philosophy of Tableaunoir (see https://github.com/tableaunoir/documentation) is KISS (keep it simple, stupid). Same for the thickness of the eraser. This is of course a matter of debate, and I am not completely convinced myself.

**color button greyed**
Indeed... but a gray button would also mean "do not click on me because I do nothing" whereas it is convenient to select a color and then automatically go back to the drawing mode and with that color.

**Rectangle eraser**
The rectangle eraser is very appealing, and also seems very useful when giving/making a presentation. You can already do that but it is too slow (tool rectangle, drag to draw the rectangle, then Ctrl+X to cut into a magnet, then Suppr to delete the magnet). A solution would be to implement: tool rectangle, drag to draw the rectangle, **then Suppr**.

What do you think?

**Reply by deleporte** [reply]
**Navigation**
Here's what I mean : the (eventual) triangle button would fold this other bar. Pressing h should hide both bars.
![image](https://user-images.githubusercontent.com/12908293/115719732-ac7bd180-a37c-11eb-9c4b-12b2727a5510.png)

**Eraser size**
I agree with the KISS, but it doesn't feel too much complicated to have the possibility to choose thickness, at least compared to the elaborated (and great!) tools for drawing circle arcs, etc.

If you don't want to add another button: maybe replace the colour button with something with two "wheels", colour and thickness?

**Rectangle eraser and magnets**
Actually it's not that slow; it's just not obvious to the newcomer.

I find it confusing that the tool for *drawing* a rectangle (or other shape) and for *selecting* a rectangle (to delete or move it) are the same. You draw a shape, then Ctrl+X, then your shape disappears and its inside has been selected... 

What about this: Ctrl+X allows to switch between drawing mode and selection/magnet mode. If you want to make a rectangular part of your screen into a magnet, first Ctrl+X (which changes how the toolbar looks: you don't want color(/thickness), pen/eraser, and text in selection mode), then draw any shape, and when you release, your shape has been selected/magnetised.

If switching to a selection mode was clearly indicated somehow (I know, I know, I keep suggesting you add buttons everywhere), then it would be clear that the correct procedure for erasing a rectangular part of the board is: switch to selection mode, select rectangular part, erase.

**Reply by francoisschwarzentruber** [reply]
I opened a new discussion about improving the toolbar (#154), because it is a different topic.
Regarding having a drawing mode and a selection mode, thank you for discussing that point. It is very important and I tend to disagree. I now explain the point of view of having a single mode (both for drawing and selection) in https://github.com/tableaunoir/documentation/blob/main/README.md
For Ctrl+X, I tend to disagree since it is the standard keyboard short-cut for cutting stuff.

I really appreciate the free discussion. Hope you are ok if I disagree with some of your points. However, it makes me really think about it and find the right justifications. So really thank you.

**Reply by deleporte** [reply]
It's okay, I'm just making suggestions based on my gut feelings, if I wanted to decide on my own what a tool does, I would develop it myself :-D 

I like your point on view on drawing/selection. Once people are accustomed to it, it's practical anyway!
